src/functional/iterator/transforms/global_tmps.py

```diff
@@ -1,25 +1,159 @@
-from typing import Dict, List
+from typing import Any, Optional
 
-from eve import NodeTranslator
-from functional.iterator import ir
+from eve import Coerced, NodeTranslator
+from eve.traits import SymbolTableTrait
+from functional.iterator import ir, type_inference
+from functional.iterator.pretty_printer import PrettyPrinter
 from functional.iterator.runtime import CartesianAxis
 from functional.iterator.transforms.collect_shifts import CollectShifts
+from functional.iterator.transforms.eta_reduction import EtaReduction
 from functional.iterator.transforms.popup_tmps import PopupTmps
+from functional.iterator.transforms.prune_closure_inputs import PruneClosureInputs
 
 
```

Comment:
Let's describe the strategy with domain update here:
- introduce magic value `_gtmp_auto_domain` in extraction pass
- then fix the domain afterwards